ykrt/src/trace/hwt/mapper.rs

```diff
@@ -1,37 +1,119 @@
 pub(crate) struct HWTTraceIterator {
-    trace: std::vec::IntoIter<TraceAction>,
+    hwt_iter: Box<dyn Iterator<Item = Result<Block, HWTracerError>> + Send>,
+    /// The next [TraceAction]`s we will produce when `next` is called. We need this intermediary
+    /// to allow us to deduplicate mapped/unmapped blocks. This will be empty on the first
+    /// iteration and from then on will always have at least one [TraceAction] in it.
+    upcoming: Vec<TraceAction>,
+    /// How many [TraceAction]s have been generated so far? We use this to know if the underlying
+    /// trace is too long.
+    tas_generated: usize,
 }
 
 impl AOTTraceIterator for HWTTraceIterator {}
 
 impl Iterator for HWTTraceIterator {
     type Item = Result<TraceAction, AOTTraceIteratorError>;
     fn next(&mut self) -> Option<Self::Item> {
-        self.trace.next().map(|x| Ok(x))
+        while self.upcoming.len() < 2 {
+            match self.hwt_iter.next() {
+                Some(Ok(x)) => {
+                    let mapped = map_block(&x);
+                    if mapped.is_empty() {
+                        // The block is unmappable. Note that we collapse consecutive unmappable
+                        // blocks.
+                        if self.upcoming.last() != Some(&TraceAction::new_unmappable_block()) {
+                            self.upcoming.push(TraceAction::new_unmappable_block());
+                            self.tas_generated += 1;
+                        }
+                    } else {
+                        for b in map_block(&x) {
+                            match b {
+                                Some(b) => {
+                                    if self.upcoming.last() != Some(&b) {
+                                        self.upcoming.push(b);
+                                        self.tas_generated += 1;
+                                    }
+                                }
+                                None => {
+                                    // Part of an hwtracer block mapped to a machine block in the
+                                    // LLVM block address map, but the machine block has no
+                                    // corresponding AOT LLVM IR blocks.
+                                    //
+                                    // FIXME: https://github.com/ykjit/yk/issues/388
+                                    // We *think* this happens because LLVM can introduce extra
+                                    // `MachineBasicBlock`s to help with laying out machine code.
+                                    // If that's the case, then for our purposes these extra blocks
+                                    // can be ignored. However, we should really investigate to be
+                                    // sure.
+                                }
+                            }
+                        }
+                    }
+                }
+                Some(Err(HWTracerError::Unrecoverable(x)))
```

Comment:
I think the above arm should be lifted out into a function. The locality of this next arm confused me for a moment: there's so much code above that I forgot that we are in a `match`.

Reply:
Keep going to a later commit and I think you'll be happy?